### PR TITLE
Frequency domain implementation

### PR DESCRIPTION
**Pull request by EthanMarx**
Adds frequency domain implementation. A bit of refactoring is done so the `TimeDomainDataset`, `FrequencyDomainDataset`, and `SpectrogramDataset` can all use `build_val_batches`. Specifically, the whitening step is now done in the subclasses, since frequency domain whitening is done differently than time domain. 

When using frequency domain dataset, there are now some redundant steps (for example, we go to the freq domain to compute SNR). This can be addressed in a separate PR, likely to ml4gw to include tools for working directly in freq domain

### CODE REVIEW

**Comment by EthanMarx** [comment]
@wbenoit26 @deepchatterjeeligo Would appreciate some sanity checking on the frequency domain implementation here whenever you get a chance.

Have a hp search running, and (it appears) we aren't going to come close to the validation score we see in time domain (~0.6). 
https://wandb.ai/ethanmarx/aframe?nw=nwuserethanmarx 

Could be a couple of reasons (e.g. ConvNet may not be as good in freq domain, etc) but curious if there's a data processing thing that I overlooked.

**Comment by EthanMarx** [comment]
Couple of notes: 

We now use `kernel_length + fduration` of fft'ed data to pass to the network. Motivation being the edges are windowed (see below) to mitigate fft artifacts.

I apply a `hann_window` to mitigate fft artifacts. It might be that this window is more aggressive than say a tukey window, and killing too much SNR.

I pass the whitened real, imaginary, abs(), and angle of the complex vector. I realize now I probably should standard scale each these since they are different magnitudes (e.g. abs() ~ real**2)

File: projects/train/train/callbacks.py [hunk 1]
```diff
@@ -57,6 +71,12 @@ def on_train_start(self, trainer, pl_module):
                     f["X"] = X.cpu().numpy()
                     f["y"] = y.cpu().numpy()
 
+                with h5py.File(
+                    os.path.join(save_dir, "val_batch.h5"), "w"
```
Comment (wbenoit26) [inline projects/train/train/callbacks.py:75]: Why save the validation batch only for local training?

File: projects/train/train/data/supervised/__init__.py [hunk 2]
```diff
@@ -1,3 +1,3 @@
-from .frequency_domain import FrequencyDomainSupervisedAframeDataset
```
Comment (wbenoit26) [inline projects/train/train/data/supervised/__init__.py:1]: Need to import `SpectrogramSuperivsedAframeDataset`. Also, I think we should either leave the `FrequencyDomainSupervisedAframeDataset` name as-is, or remove `Supervised` from the other class names to maintain consistency. I'm good with either option.

File: projects/train/train/data/supervised/frequency_domain.py [hunk 3]
```diff
@@ -23,14 +23,73 @@ def setup(self, *args, **kwargs):
...
+
+        # build tapering window
+        window_length = self.hparams.kernel_length + self.hparams.fduration
+        window_size = int(window_length * self.sample_rate)
+        window = torch.hann_window(window_size)
+        df = 1 / window_length
+        window_scale = torch.sum(window**2) / len(window)
+        window /= torch.sqrt(window_scale)
+        window *= torch.sqrt(torch.tensor(4 * df))
+        self.window = window[None][None].to(self.device)
```
Comment (wbenoit26) [inline projects/train/train/data/supervised/frequency_domain.py:62]: Can you explain or point to something that explains why the window is constructed this way?

File: projects/train/train/data/supervised/frequency_domain.py [hunk 4]
```diff
@@ -23,14 +23,73 @@ def setup(self, *args, **kwargs):
...
+
+    def build_val_batches(self, *args, **kwargs):
+        X_bg, X_inj, psds = super().build_val_batches(*args, **kwargs)
+
+        # fft and whiten in frequeny domain
```
Comment (wbenoit26) [inline projects/train/train/data/supervised/frequency_domain.py:67]: Typo in frequency

**Review by wbenoit26 — COMMENTED** [review]
Left a few comments on minor things, but I don't see anything that you haven't already suggested. The scale between the components and the magnitude should be similar after whitening, so I don't think that's the problem, but doesn't hurt to scale. My money is on the architecture being less effective for this representation.

One fairly quick test would be to apply the window to the time-domain network and see if the validation score falls off in the same way there.

**Comment by EthanMarx** [comment]
@wbenoit26 Made a couple other clean up changes you might want to look at.

File: projects/train/train/data/supervised/frequency_domain.py (hunk 3 above)
Comment (EthanMarx) [reply projects/train/train/data/supervised/frequency_domain.py:86]: Yeah this is to ensure that the fft has unit variance mean 0 after the fft - the window takes away some power, so you have to divide by this window factor

**Comment by EthanMarx** [comment]
And FWIW did a test run with the `TimeDomainDataset` and original training behavior is reproduced over ~20 epochs

File: projects/train/train/data/supervised/frequency_domain.py [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import torch
 
+from ml4gw.spectral import truncate_inverse_power_spectrum
 from ml4gw.transforms.qtransform import SingleQTransform
 from train.data.base import Tensor
 from train.data.supervised.supervised import SupervisedAframeDataset
 
 
-class SpectrogramSupervisedAframeDataset(SupervisedAframeDataset):
+class SpectrogramDomainSupervisedAframeDataset(SupervisedAframeDataset):
```
Comment (wbenoit26) [inline projects/train/train/data/supervised/frequency_domain.py:9]: Not relevant to this PR, but we could consider removing `Aframe` from all these names to make them a little more manageable

Comment (EthanMarx) [reply projects/train/train/data/supervised/frequency_domain.py:9]: Good idea

File: .test [hunk 6]
```diff
@@ -0,0 +1,7 @@
+AFRAME_BASE=~/aframe # base location for aframe runs
```
Comment (wbenoit26) [inline .test:1]: Extra file

**Comment by wbenoit26** [comment]
Have you started a training with this setup? Just using the real and imaginary parts, I mean

**Comment by EthanMarx** [comment]
No, will launch a run soon trying out the best hps from that search. Will eventually re run a search with just real and imaginary / try out a dense net

**Pull request merged**